docs/player.rst

```diff
@@ -0,0 +1,143 @@
+==============
+
+The :py:class:`~.GameSession` client API closely mimics how an individual player
+would operate and which information they would need during a session.
+Thus, for client programs that already have a ``class Player`` abstraction,
+the client API can be integrated via the `Adapter Pattern`_ by implementing
+a new kind of ``Player`` as an adapter to the :py:class:`~.GameSession`.
+
+.. note::
+
+    It is important to distinguish between "the player" as a human being
+    *using* the client program and "the ``Player``" as an abstraction as
+    *part of*  the client program.
+    Keep in mind that "the ``Player``" needs access to more information
+    than "the player";
+    this becomes apparent when wrapping a :py:class:`~.GameSession`,
+    since it means explicitly giving "the ``Player``" information
+    about its opponent.
+
+Since the adapter ``Player`` needs access to information that is usually
+communicated implicitly – for example, the ship positions shown via the UI
+or even the length of a session – it is very likely necessary to extend the
+``Player`` abstraction to match.
+The following is an outline of how such an adapter can be designed.
+
+For simplicity, this outline only assumes
+one "local" player directly using the client program, and
+one "remote" player connected via the :py:class:`~.GameSession`.
+You likely still want to support local vs local play,
+and might want to support remote vs remote play as well.
+
+Creating Instances
+------------------
+
+A :py:class:`~.GameSession` needs to be scoped to one round of the game
+and can only disclose the opponent player name after connecting to a game.
+This prevents just constructing a ``RemotePlayer`` that wraps the session.
+
+We recommend to extend the ``Player`` abstraction with:
+
+- an alternative constructor via :py:func:`classmethod` which is
+- scoped to one round of the game via :py:func:`~contextlib.contextmanager`.
+
+.. code:: python3
+
+    class RemotePlayer:
+        # this constructor usually is not called manually:
+        # 'identifier' is provided by the player on the remote session!
+        def __init__(self, identifier: str, session: GameSession):
+            self.identifier = identifier
+            self._session = session
+
+        @classmethod
+        @contextmanager
+        def create(cls, opponent: "str | None"):
+            """
+            Create a scoped instance of this class and inform it of its ``opponent``
+            """
+            # 'opponent' is the local player, 'session.opponent' the remote player
+            with GameSession.connect(opponent) as session:
+                yield cls(session.opponent, session)
+
+The constructor can then be used in a `with statement`_:
+
+.. code:: python3
+
+    with RemotePlayer.create(...) as player:
+        ...
+
+Note that you can create two players of type ``P1`` and ``P2``
+in one ``with`` statement using
+``with P1.create(...) as player_1, P2.create(...) as player_1:``.
```

Comment:
Shouldn't it be `with P1.create(...) as player_1, P2.create(...) as player_2`?

Reply:
Good catch, the numbering is consistent now.